keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2314 @@
+are not available in Kubernetes anymore. The resources are still supported by
+the underlying runtime but disabling the feature in Kubernetes makes them
+unavailable and the related PodSpec fields are not accepted in validation.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Workloads might have failed because of unsupported fields in the pod spec and
+need to be restarted.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+Kubelet unit tests are extended to verify that no QoS-class resource
+assignments are correctly passed down to the CRI API, also verifying that
+assignments are not passed down if the feature gate is disabled.
+
+Apiserver unit tests are extended to verify that the new fields in PodSpec are
+preserved over updates of the Pod object, even when the feature gate is
+disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+Implementation Phase 1: Already running workloads ahouls not be affected as the
+QoS-class resources feature operates on new fields in the PodSpec. Bugs in
+kubelet might cause containers fail to start, either by failing a pod admission
+check that should pass or passing incorrect parameters to the container
+runtime. Bugs in kube-scheduler might leave Pods in pending state (even though
+they could be run in some node) or scheduling on an incorrect node, causing
+kubelet todeny running the pod.
+
+Future implementation phases: TBD.
+
+###### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+Implementation Phase 1: Watch for non-ready pods with CreateContainerError
```

Comment:
We should probably add a metric to scheduler for this.

Reply:
The CreateContainerError originates from the CRI runtime and has nothing to do with scheduling. There's already e.g. `kubelet_started_containers_errors_total` metric in the kubelet. That if we state that for kube-scheduler and kubelet the existing metrics can be used, WDYT @logicalhan?

Reply:
Okay that sounds reasonable to me.

Reply:
OK, added sentence `For kube-scheduler and kubelet the existing metrics (e.g. kubelet_started_containers_errors_total) can be used.`